internal/terraform/node_output.go

```diff
@@ -781,6 +781,14 @@ func (n *NodeApplyableOutput) setValue(namedVals *namedvals.State, state *states
 		// not serialized.
 		if n.Addr.Module.IsRoot() {
 			val, _ = val.UnmarkDeep()
+			if val.IsKnown() && !val.IsWhollyKnown() {
+				// For the version of the value that ends up in state, treat
+				// partially-unknown values as wholly unknown. The goal is to avoid
+				// confusion with outputs that reference deferred values; since the
+				// referenced resource is still unready, we don't want to store
+				// planned-but-not-verified speculations in the source of truth.
+				val = cty.UnknownVal(val.Type())
+			}
 			val = cty.UnknownAsNull(val)
```

Comment:
```suggestion
			if val.IsKnown() && !val.IsWhollyKnown() {
				// For the version of the value that ends up in state, treat
				// partially-unknown values as wholly unknown. The goal is to avoid
				// confusion with outputs that reference deferred values; since the
				// referenced resource is still unready, we don't want to store
				// planned-but-not-verified speculations in the source of truth.
				val = cty.NullVal(val.Type())
			} else {
			        val = cty.UnknownAsNull(val)
			}
```

I think the follow up call to UnknownAsNull will just always replace the unknown value we just created with null? Which seems a bit wasteful when we can just set the thing to null directly. What do you think? Probably need to update the comment a bit to reflect we're not setting it to unknown any more.

Reply:
Thanks for catching this — I agree. 👍🏼 Will change.